File: leet_code/79_word_search.js

/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function(board, word) {

    const ROWS = board.length;
    const COLS = board[0].length;

    function rec(row, col, idx, paths=null){

        if (paths===null){
            paths = new Set();
        }

        if (idx === word.length){
            return true;
        }
        if (
            paths.has(`${row},${col}`) ||
            row < 0 ||
            col < 0 ||
            row >= ROWS ||
            col >= COLS
        ){
            return false;
        }
        if (word[idx] == board[row][col]){
            paths.add(`${row},${col}`)
            res = (
                rec(row+1, col, idx+1, paths)  ||
                rec(row-1, col, idx+1, paths)  ||
                rec(row, col+1, idx+1, paths)  ||
                rec(row, col-1, idx+1, paths) 
            )
            paths.delete(`${row},${col}`)
            return res
        }
        return false
    }
    
    for (let row=0; row < ROWS; row++){
        for(let col=0; col < COLS; col++){
            if (rec(row, col, 0)){
                return true;
            }
        }
    }
    return false
};







const board = [
    Array.from("SIEE"),
    Array.from("SFMS"),
    Array.from("ONAE")
  ];
  const words = ["SIEMANO", "SIEMA", "SIEMU"];
  

  // Print the board
  board.forEach(row => {
    console.log(row.join(' '));
  });
  
  // Print the words and whether they exist on the board
  words.forEach(word => {
    console.log("Word:", word);
    // Call the exist method and print the result (implement the exist method within the Solution class)
    console.log(`word_exist ${exist(board, word)}`);
  });

File: leetcode/79_word_search.py

from typing import List


# TC O(row*col*4^len(word)) SC(len(word))
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        ROWS = len(board)
        COLS = len(board[0])
        def dfs(row, col, word_idx, paths= None) -> bool:
            if paths is None:
                paths = set()
            # base cases
            if word_idx == len(word):
                return True
            false_cond = (
                row < 0 or
                col < 0 or
                row >= ROWS or
                col >= COLS or
                (row, col) in paths
            )
            if false_cond:
                return False
            
            # recursive case
            if word[word_idx] == board[row][col]:
                paths.add((row, col))
                res = (
                    dfs(row+1, col, word_idx+1, paths) or
                    dfs(row-1, col, word_idx+1, paths) or
                    dfs(row, col+1, word_idx+1, paths) or
                    dfs(row, col-1, word_idx+1, paths) 
                )
                paths.remove((row, col))
                return res
            return False
        
        for row in range(ROWS):
            for col in range(COLS):
                if dfs(row, col, 0):
                    return True
        return False


board = [
    list("SIEE"),
    list("SFMS"),
    list("ONAE")
]
word = "SIEMANO"

s = Solution()

# Print the board
for row in board:
    print(' '.join(row))


word = "SIEMANO"
# Print the word
print("Word:", word)
print(f"word_exist {s.exist(board, word)}")

word = "SIEMA"
# Print the word
print("Word:", word)
print(f"word_exist {s.exist(board, word)}")

word = "SIEMU"
# Print the word
print("Word:", word)
print(f"word_exist {s.exist(board, word)}")